src/Xamarin.Android.Build.Tasks/Utilities/AssemblyCompression.cs

```diff
@@ -0,0 +1,83 @@
+using System;
+using System.Buffers;
+using System.IO;
+
+using K4os.Compression.LZ4;
+
+namespace Xamarin.Android.Tasks
+{
+	class AssemblyCompression
+	{
+		public sealed class AssemblyData
+		{
+			public string SourcePath { get; internal set; }
+			public uint DescriptorIndex { get; internal set; }
+
+			public string DestinationPath;
+			public uint SourceSize;
+			public uint DestinationSize;
+
+			public AssemblyData (string sourcePath, uint descriptorIndex)
+			{
+				SetData (sourcePath, descriptorIndex);
+			}
+
+			public void SetData (string sourcePath, uint descriptorIndex)
+			{
+				if (String.IsNullOrEmpty (sourcePath))
+					throw new ArgumentException ("must not be null or empty", nameof (sourcePath));
+				SourcePath = sourcePath;
+				DescriptorIndex = descriptorIndex;
+			}
+		}
+
+		const uint CompressedDataMagic = 0x5A4C4158; // 'XALZ', little-endian
+
+		static readonly ArrayPool<byte> bytePool = ArrayPool<byte>.Create ();
+
+		public static bool Compress (AssemblyData data)
+		{
+			if (data == null)
+				throw new ArgumentNullException (nameof (data));
+
+			var fi = new FileInfo (data.SourcePath);
+			if (!fi.Exists)
+				throw new InvalidOperationException ($"File '{data.SourcePath}' does not exist");
+			data.DestinationPath = $"{data.SourcePath}.lz4";
+			data.SourceSize = (uint)fi.Length;
+
+			byte[] sourceBytes = null;
+			byte[] destBytes = null;
+			try {
+				sourceBytes = bytePool.Rent ((int)fi.Length);
+				using (var fs = File.Open (data.SourcePath, FileMode.Open, FileAccess.Read, FileShare.Read)) {
+					fs.Read (sourceBytes, 0, (int)fi.Length);
+				}
```

Comment:
This loads the entire file in one huge `byte[]`, could that break if you had a huge assembly? I'm remembering the person that had a huge `.mp4` file as an `EmbeddedResource`. 👀 

Should this use a smaller buffer, like `8 * 1024` and do the `LZ4Codec.Encode` calls in a loop?

Reply:
There is a separate package `K4os.Compression.LZ4.Streams` which is designed to work efficiently with streaming stuff, perhaps we should look at using that for this section.

Reply:
Stream docs https://github.com/MiloszKrajewski/K4os.Compression.LZ4#stream-compression

Reply:
I will need to address it somehow, but I don't want to use the `Streams` package for LZ4 - it generates the lz4 frame format output and what I'm using here is the block ("raw" - no metadata) compression. I need this to keep the runtime side fast (decompressing lz4 frames requires a bit of housekeeping I want to avoid)

Reply:
@jonathanpeppers alas, doing that in a loop is not possible with `K4os.Compression.LZ4`, not with its public interfaces. Doing that with what we have available would result in X separately compressed blocks, thus complicating handling of that data during the runtime (including a slowdown as each block would require a new setup of the decoding engine) and hurting compression. Internally, `K4os.Compression.LZ4` appears to have a way to do compression the way you suggest, I'll take a look into it later on.

Reply:
I think we should leave this as it is. After I thought about this more the max amount of memory will be the size of the largest assembly + a bit for the compression buffer. This is because we are recycling the buffers. I think it would be a problem if we constantly allocated new buffers. 

Lets go with this for now and alter it if customers start hitting issues.